XMonad/Util/OneState.hs

```diff
@@ -0,0 +1,142 @@
+{-# LANGUAGE AllowAmbiguousTypes        #-}
+{-# LANGUAGE DerivingStrategies         #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE LambdaCase                 #-}
+{-# LANGUAGE ScopedTypeVariables        #-}
+{-# LANGUAGE TypeApplications           #-}
+{-# LANGUAGE TypeFamilies               #-}
+{-# OPTIONS_GHC -Wall -Werror #-}
+
+module XMonad.Util.OneState
```

Comment:
One quick thing: the implementation of this is pretty neat, the only question I have is whether we actually want to unify ExtensibleState and ExtensibleConf into one thing.  Personally, I find the current situation quite convenient in terms of having a mental model of what a given piece of code can do, but maybe I'm alone in that (Cc. @liskin as the author of ExtensibleConf)

Reply:
I don't think it's much about the situation being convenient or not. It's just a very elaborate workaround for not being able to initialise/modify (`put`) ExtensibleState in config-time. Or, to be more precise, not having a nice interface for it, as one can indeed do it using `startupHook`, but that's ugly, and composes poorly.

(I have yet to take a look at the rest of the code so I don't really have an opinion whether this elaborate workaround is worth it or whether there are easier ways to solve the same problem.)

Reply:
I believe `OneState` was created for two reasons. (1) I found combining state and config to be conceptually simpler and provide more flexibility for free (and at the time I wasn't too focused on thinking about merging into `xmonad-contrib`); and (2) I wanted to allow modifying the config at runtime, similar to what @liskin is saying.

Currently I don't think (2) is actually used. Originally I wanted to use it in order to allow adjustments to the grid layout at runtime, so you could, for instance, add and remove rows on-demand. I still think this would be worthwhile, although it's not on my TODO list at this moment.

Reply:
Do we want the config to be modifiable at runtime though? I'm not sure

> as one can indeed do it using startupHook, but that's ugly, and composes poorly.

@liskin could you elaborate? I mean, we can have a nice interface where we internally collect all defaults from the config, and compose then apply them in the startupHook. Might need some type magic, but I don't see why that's a problem.

Reply:
Yeah, we can have an interface, and `OneState` is one such interface. Although it doesn't use `startupHook`, but there's a good reason for it: it really does compose poorly. You can't guarantee that a specific part of `startupHook` runs before everything else, so if you want to make a general-purpose ExtensibleState-like interface that can be initialised in config-time, you need to make it work even if the hook hasn't run yet. So you need to do exactly what `OneState` does: look into ExtensibleState and fall back to ExtensibleConf, every single time you access it.

If you didn't need a general-purpose interface and you knew the order of `startupHooks` doesn't matter in your specific case, then it's okay to just do a single ExtensibleConf → ExtensibleState sync. It's not safe to do this in a general-purpose interface though.

Anyway, looks like nobody really needs this functionality now… :-)

Reply:
> Anyway, looks like nobody really needs this functionality now… :-)

Well, I would like to export an all-powerful `(State -> State) -> X ()` function from `Grid` so that users may modify their grid layout at runtime if they would like. (While one can certainly do without such a capability, I think it's also a reasonable desire)